dataRetriever/dataPool/nonceSyncMapCacher.go

```diff
@@ -0,0 +1,151 @@
+package dataPool
+
+import (
+	"bytes"
+	"sync"
+
+	"github.com/ElrondNetwork/elrond-go/core/logger"
+	"github.com/ElrondNetwork/elrond-go/data/typeConverters"
+	"github.com/ElrondNetwork/elrond-go/dataRetriever"
+	"github.com/ElrondNetwork/elrond-go/storage"
+)
+
+var log = logger.DefaultLogger()
+
+type nonceSyncMapCacher struct {
+	mergeMut             sync.Mutex
+	cacher               storage.Cacher
+	nonceConverter       typeConverters.Uint64ByteSliceConverter
+	mutAddedDataHandlers sync.RWMutex
+	addedDataHandlers    []func(nonce uint64, shardId uint32, value []byte)
+}
+
+// NewNonceSyncMapCacher returns a new instance of nonceSyncMapCacher
+func NewNonceSyncMapCacher(
+	cacher storage.Cacher,
+	nonceConverter typeConverters.Uint64ByteSliceConverter,
+) (*nonceSyncMapCacher, error) {
+
+	if cacher == nil {
+		return nil, dataRetriever.ErrNilCacher
+	}
+	if nonceConverter == nil {
+		return nil, dataRetriever.ErrNilUint64ByteSliceConverter
+	}
+
+	return &nonceSyncMapCacher{
+		cacher:            cacher,
+		nonceConverter:    nonceConverter,
+		addedDataHandlers: make([]func(nonce uint64, shardId uint32, value []byte), 0),
+	}, nil
+}
+
+// Clear is used to completely clear the cache.
+func (nspc *nonceSyncMapCacher) Clear() {
+	nspc.cacher.Clear()
+}
+
+// Get looks up for a nonce in cache.
+func (nspc *nonceSyncMapCacher) Get(nonce uint64) (dataRetriever.ShardIdHashMap, bool) {
+	val, ok := nspc.cacher.Peek(nspc.nonceConverter.ToByteSlice(nonce))
+	if !ok {
+		return nil, ok
+	}
+
+	syncMap, ok := val.(*ShardIdHashSyncMap)
+	if !ok {
+		return nil, ok
+	}
+
+	return syncMap, ok
+}
+
+// Merge will append existing values from src map. If the keys already exists in the existing map, their values
+// will be overwritten. If the existing map is nil, a new map will created and all values from src map will be copied.
+func (nspc *nonceSyncMapCacher) Merge(nonce uint64, src dataRetriever.ShardIdHashMap) {
+	if src == nil {
+		return
+	}
+
+	nspc.mergeMut.Lock()
+	defer nspc.mergeMut.Unlock()
+
+	shouldRewriteMap := false
+	val, ok := nspc.cacher.Peek(nspc.nonceConverter.ToByteSlice(nonce))
+	if !ok {
+		val = &ShardIdHashSyncMap{}
+		shouldRewriteMap = true
+	}
+
+	syncMap := val.(*ShardIdHashSyncMap)
+
+	if shouldRewriteMap {
+		nspc.cacher.Put(nspc.nonceConverter.ToByteSlice(nonce), syncMap)
+	}
+
+	nspc.copySyncMap(nonce, syncMap, src)
+}
+
+func (nspc *nonceSyncMapCacher) copySyncMap(nonce uint64, dest dataRetriever.ShardIdHashMap, src dataRetriever.ShardIdHashMap) {
+	src.Range(func(shardId uint32, hash []byte) bool {
+		existingVal, exists := dest.Load(shardId)
+		if !exists {
+			//new key with value
+			dest.Store(shardId, hash)
+			nspc.callAddedDataHandlers(nonce, shardId, hash)
+			return true
+		}
+
+		if !bytes.Equal(existingVal, hash) {
+			//value mismatch
+			dest.Store(shardId, hash)
+			nspc.callAddedDataHandlers(nonce, shardId, hash)
+			return true
+		}
+
+		return true
```

Comment:
True signals that the range func can move to the next value from the map. Since we had to satisfy the func signature, we had to return true as to try each and every value from the map.